Support generating sourceMaps for source debugging

aviettran opened the pull request:
**Justification**

Sucrase, by default, does not generate source maps. In order for IDEs (in my case, vscode) to allow for breakpoints on source files (\*.ts), [source maps must exist](https://code.visualstudio.com/docs/nodejs/nodejs-debugging). The only alternative, which is quite tedious, is to find the built Javascript (\*-dist/\*\*/\*) corresponding to the source file (\*.ts) and to create the corresponding breakpoint there.

**Proposed Solution**

Sucrase does not allow its cli tool to generate source maps (https://github.com/alangpierce/sucrase/issues/223), nor does it seem to be a priority. It must be done programmatically. The proposed solution is to convert the cli-based transcoding process into a programmatic one ~~which, optionally,~~ generates source maps.

~~**Proposed Configuration**
Use the node module dotenv to allow the developer to create a .env file in order to set the environment variable "NODE_ENV" to "development". This will enable the generation of sourceMaps. This was just a standard and quick way for me to get this PR up and running. I am very open to alternative suggestions.~~

**Proposed Build-Artifacts**

For each output JavaScript file (\*.js), generate a corresponding \*.js.map file and place it in a new directory called "sourceMaps".

**Example:**

**Source:**
data/moves.ts

**Output:**
.data-dist/moves.js
.data-dist/sourceMaps/moves.js.map

In addition to these new \*.map files, each output JavaScript file will have a newly appended comment at the end of the file:
 //# sourceMappingURL=sourceMaps/moves.js.map

~~CI pipelines and production builds _should not_ have the environment "NODE_ENV" set to "development" and will, therefor, not have any change in behavior as a result of this PR.~~

AnnikaCodes commented on build:
```diff
@@ -8,10 +9,13 @@ try {
 }
 
-var child_process = require('child_process');
-var fs = require('fs');
-var path = require('path');
-var force = process.argv[2] === '--force';
```
IIRC these are intended to be `var`, not `const`, so that really old node versions will reach a point where a more descriptive error can be thrown.

AnnikaCodes replied:
I don't really think that decision still makes sense several years later, but that's a separate change.

Zarel replied:
Yeah, we should probably put the explanation in comments at the top of the file, so this doesn't keep happening.

aviettran replied:
Noted.

aviettran replied:
I reverted any eslint auto-fixed changes and added exceptions to `.eslintrc-no-types.json`.

Zarel commented:
How do source maps impact performance? If it's light enough, we might prefer to always enable source maps. In general, we prefer not to have differences between development and production builds (so that debugging issues in prod isn't harder than it needs to be).

Zarel commented:
(That would also sidestep the `dotenv` dependency which I'm not a big fan of adding.)

aviettran commented:
Build times are not significantly impacted:

aviettran:~/pokemon-showdown$ time ./build --force

real    0m6.779s
user    0m6.031s
sys     0m1.375s
aviettran:~/pokemon-showdown$ export NODE_ENV=development
aviettran:~/pokemon-showdown$ time ./build --force

real    0m7.071s
user    0m5.922s
sys     0m1.500s

Zarel commented:
Indeed. Would you be willing to remove all the env stuff, then? Just build source maps all the time?

urkerab commented on .gitignore:
```diff
@@ -16,10 +16,12 @@ package-lock.json
 .*-dist/
 tools/set-import/importer.js
 tools/modlog/converter.js
+**/sourceMaps
```
Isn't this already covered by the `.*-dist` exclusion?

aviettran replied:
`tools/set-import/importer.js` and `converter.js` are listed explicitly, so rather than repeat them, I just added a rule for **/sourceMaps.

aviettran commented:
@Zarel I removed anything conditional from the `sourceMap` generation logic and removed `dotenv`. I also removed the `sourceRoot` property from `*.map`, since it can hypothetically get deployed somewhere other than where it was built. The relative directory structure will have to stay intact for maps to work, though.

Zarel commented on .eslintrc-no-types.json:
```diff
@@ -307,6 +307,13 @@
+    {
+      "files": ["build"],
+      "rules": {
+        "strict": "off",
```
`"use strict"` is fine, though, right? No compatibility issues there.

Zarel commented on build:
```diff
@@ -8,6 +8,7 @@ try {
 	process.exit(1);
 }
 
+var {transform} = require('sucrase');
```
```suggestion
var transform = require('sucrase').transform;
```

(Destructuring is also not supported by early Node versions)

aviettran replied:
Ok. What is the minimum version we are supporting? Let me know so I can test.

Zarel commented on build:
```diff
@@ -101,13 +102,74 @@ function needsSucrase(source, dest, path = '') {
+		const outChildPath = path.join(outDirPath, child);
+		const pathToSourceChild = path.join(pathToSource, '..');
+		if (fs.statSync(srcChildPath).isDirectory()) {
+			const innerOptions = {...options};
```
```suggestion
			const innerOptions = Object.assign({}, options);
```

`Object.assign` also postdates early Node, but it's not syntactical so we'd hit the error before hitting this.

Honestly, if we want to write modern code here, we might want to put it in a different file wrapping sucrase, rather than in this file with a lot of restrictions on what syntax is allowed.

aviettran replied:
I took your suggestion and moved a couple functions out. Let me know if it works, or if you want me to put the lib somewhere else.

Zarel commented on build:
```diff
@@ -101,13 +102,74 @@ function needsSucrase(source, dest, path = '') {
+			innerOptions.outDirPath = outChildPath;
+			innerOptions.pathToSource = pathToSourceChild;
+			const innerFiles = findFiles(innerOptions);
+			outArr.push(...innerFiles);
```
```suggestion
			Array.prototype.push.apply(outArr, innerFiles);
```

Zarel commented on build:
```diff
@@ -101,13 +102,74 @@ function needsSucrase(source, dest, path = '') {
+		} else if (extensions.some((ext) => srcChildPath.endsWith(ext))) {
+			const outPath = outChildPath.replace(
+				/\.\w+$/,
+				`.${options.outExtension}`
```
```suggestion
				"." + options.outExtension
```

Zarel commented on build:
```diff
@@ -101,13 +102,74 @@ function needsSucrase(source, dest, path = '') {
+			innerOptions.pathToSource = pathToSourceChild;
+			const innerFiles = findFiles(innerOptions);
+			outArr.push(...innerFiles);
+		} else if (extensions.some((ext) => srcChildPath.endsWith(ext))) {
```
```suggestion
		} else if (extensions.some(function (ext) { return srcChildPath.endsWith(ext); })) {
```

Zarel commented on build:
```diff
@@ -101,13 +102,74 @@ function needsSucrase(source, dest, path = '') {
+	const sucraseOptions = {transforms: ['typescript', 'imports'], enableLegacyTypeScriptModuleInterop: true, ...opts};
+	const files = findFiles({outDirPath: out, srcDirPath: src, sucraseOptions: sucraseOptions, excludeDirs: ['sourceMaps', ...excludeDirs], outExtension: 'js'});
+	for (const file of files) {
+		const sucraseOptionsFile = {...sucraseOptions, sourceMapOptions: {compiledFilename: file.outPath}, filePath: path.join(file.pathToSource, file.srcPath)};
```
```suggestion
		const sucraseOptionsFile = Object.assign(
			{},
			sucraseOptions,
			{sourceMapOptions: {compiledFilename: file.outPath}, filePath: path.join(file.pathToSource, file.srcPath)}
		);
```

Zarel commented:
Looks great! Thanks!
